### PR TITLE
hero.detach

### PR DESCRIPTION
**Pull request by calebjclark**
This PR adds documentation that was missing from the previous "commit" PR.

It also implements a number of changes to the "commit" PR in order to simplify concepts/usage.

- HeroExtractor has been renamed to HeroReplay.
- The previous concept of FrozenDOM has been renamed to DetachedDom.
- element.$extract has been removed.
- element.$collect has been renamed to element.$detach
- hero.collect has been split into two methods:
  - hero.detach allows detaching elements and resources.
  - hero.setData allows saving arbitrary JSON-able data to the session db
- CollectedElements has been renamed to DetachedElements
- CollectedResources has been renamed to DetachedResources
- CollectedSnippets has been replaced with hero.getData

### CODE REVIEW

File: docs/main/BasicClient/Hero.md [hunk 1]
```diff
@@ -257,6 +255,33 @@ Close a single Tab. The first opened Tab will become the focused tab.
...
+
+### hero.detach *(elementOrResource, options?)* {#detach}
+
+Detaches an element or resource from AwaitedDOM and converts it into DetachedDOM. You can supply an optional name in the 2nd argument to store the element in [hero.detachedElements](/docs/hero/basic-client/hero#detachedElements) for later use.
+
+```js
+const hero = new Hero();
+await hero.goto('https://ulixee.org');
+await hero.detach('title', hero.querySelector('h1'));
+const when = await hero.detachedElements.get('title');
```
Comment (blakebyrnes) [inline docs/main/BasicClient/Hero.md:267]: Copy paste error with the name here

File: client/lib/DomExtender.ts [hunk 2]
```diff
@@ -25,7 +25,8 @@ import Interactor from './Interactor';
 import { getAwaitedPathAsMethodArg } from './SetupAwaitedHandler';
 import { scriptInstance } from './internal';
 import { IExtractElementFn, IExtractElementOptions, IExtractElementsFn } from '../interfaces/IExtractElementFn';
-import CollectedElements from './CollectedElements';
+import DetachedElements from './DetachedElements';
+import DetachedDOM from './DetachedDOM';
```
Comment (blakebyrnes) [inline client/lib/DomExtender.ts:29]: Lint error here, and there's an unused function in this class too

File: docs/main/BasicClient/AwaitedDOM-Extensions.md [hunk 3]
```diff
@@ -106,7 +93,35 @@ Attached to Nodes and Elements ([see list](#super-nodes)).
 
 #### **Returns**: `Promise<void>`
 
-### node.$type *(...typeInteractions)* {#type}
+
+### element.$detach *(name?)* {#extract-later}
+
+Detaches an element from the browser's live [AwaitedDOM](/docs/hero/basic-client/awaited-dom) environment and converts it into a [DetachedDOM](/docs/hero/basic-client/detached-dom) object for local usage (i.e, without any need for promises or awaits). You can optionally supply a name as the first argument if you want to add it to the [hero.detachedElements](/docs/hero/basic-client/hero#detached-elements) bucket. The advantage of hero.detachedElements is you can use and reuse them from within [HeroReplay(/docs/hero/basic-client/hero-replay) long after your Hero session has closed. This allows you to write extraction logic that can be easily iterated on without needing to reload the webpage(s).
```
Comment (blakebyrnes) [inline docs/main/BasicClient/AwaitedDOM-Extensions.md:99]: Broken link to HeroReplay

File: docs/main/BasicClient/AwaitedDOM-Extensions.md [hunk 4]
```diff
@@ -106,7 +93,35 @@ Attached to Nodes and Elements ([see list](#super-nodes)).
...
+```
+
+You can create a second script that uses [HeroReplay](/docs/hero/basic-client/hero-replay) to find the data you need without loading the website again:
+```js
+const hero = new HeroReplay({ /* previousSessionId */});
+const h1 = await hero.detachedElements.get('title');
+const h1Children = [...h1.querySelectorAll('div')].map(x => x.textContent);
+```
+
+It's important to remember that the returned element returned is no longer [AwaitedDOM](/docs/hero/basic-client/awaited-dom). It is in the form of [DetachedDOM](/docs/hero/basic-client/detached-dom), which allows access to properties and methods without the `await` keyword.
```
Comment (blakebyrnes) [inline docs/main/BasicClient/AwaitedDOM-Extensions.md:116]: Why is it important? Doesn't it work with or without the awaits?

File: docs/main/BasicClient/AwaitedDOM-Extensions.md [hunk 5]
```diff
@@ -106,7 +93,35 @@ Attached to Nodes and Elements ([see list](#super-nodes)).
...
+
+#### **Arguments**:
+
+- name `string`. Optional. The name used to retrieve this element from from [hero.detachedElements](/docs/databox/basic-client/hero#detached-elements). Leaving out this argument ensures the element is not added to detachedElements.
```
Comment (blakebyrnes) [inline docs/main/BasicClient/AwaitedDOM-Extensions.md:120]: `from from`

File: docs/main/BasicClient/Hero.md [hunk 6]
```diff
@@ -257,6 +255,33 @@ Close a single Tab. The first opened Tab will become the focused tab.
 
 Alias for [Tab.close()](/docs/hero/basic-client/tab#close)
 
+
+### hero.detach *(elementOrResource, options?)* {#detach}
+
+Detaches an element or resource from AwaitedDOM and converts it into DetachedDOM. You can supply an optional name in the 2nd argument to store the element in [hero.detachedElements](/docs/hero/basic-client/hero#detachedElements) for later use.
```
Comment (blakebyrnes) [inline docs/main/BasicClient/Hero.md:261]: Is a resource a DetachedDOM? Or is this description just missing something?

File: docs/main/BasicClient/Hero.md (hunk 1 above)
Comment (calebjclark) [reply docs/main/BasicClient/Hero.md:267]: What is the error?

Comment (blakebyrnes) [reply docs/main/BasicClient/Hero.md:267]: No error, but the name of "when" seemed like it was copied from the example where you saved "time". It threw me off here.

File: docs/main/BasicClient/FlowHandling.md [hunk 7]
```diff
@@ -10,12 +10,12 @@ Often while building a scraper script, the script is built to handle the "clean
 To that end, FlowHandlers were created. FlowHandlers allow you to define a "State" that the page can be in, and a "Resolver" to correct the state and resume your "clean path". Because these are outlier paths, the expectation is that FlowHandlers will click the consent button and resume, or redirect back to the page you wish to be on (if you've been sent to another page).
 
 ```js
-5.   await hero.registerFlowHandler('CookieModal', assert => {
-6.     assert(hero.querySelector('#cookie-modal').$isVisible);
-7.   },
-8.   async error => {
-9.     await hero.querySelector('#cookie-modal .dismiss').$click();
-10.  });
+await hero.registerFlowHandler('CookieModal', assert => {
```
[the remark this answers is not in the document]
Comment (calebjclark) [reply docs/main/BasicClient/FlowHandling.md:13]: Line numbers were no longer be referenced.

File: docs/main/BasicClient/Hero.md (hunk 6 above)
Comment (calebjclark) [reply docs/main/BasicClient/Hero.md]: I added clarification that it is a DetachedResource.

File: client/lib/DomExtender.ts (hunk 2 above)
Comment (calebjclark) [reply client/lib/DomExtender.ts]: Fixed

File: docs/main/BasicClient/AwaitedDOM-Extensions.md (hunk 3 above)
Comment (calebjclark) [reply docs/main/BasicClient/AwaitedDOM-Extensions.md]: Fixed

File: docs/main/BasicClient/AwaitedDOM-Extensions.md (hunk 4 above)
Comment (calebjclark) [reply docs/main/BasicClient/AwaitedDOM-Extensions.md]: Yeah, the wording was too much. I simplified and removed the "It's important".

File: docs/main/BasicClient/AwaitedDOM-Extensions.md (hunk 5 above)
Comment (calebjclark) [reply docs/main/BasicClient/AwaitedDOM-Extensions.md]: Fixed

File: docs/main/BasicClient/AwaitedDOM-Extensions.md [hunk 8]
```diff
@@ -202,20 +217,42 @@ Attached to Nodes and Elements ([see list](#super-nodes)).
 
 #### **Returns**: `Promise<ISuperElement>`
 
+
 ## Collection Methods
```
[the remark this answers is not in the document]
Comment (calebjclark) [reply docs/main/BasicClient/AwaitedDOM-Extensions.md:223]: Done.

File: docs/main/BasicClient/DetachedDOM.md [hunk 9]
```diff
@@ -20,7 +22,7 @@ const fs = require('fs');
 const { DetachedDOM } = require('@ulixee/hero');
 
 const html = fs.readFileSync('/docs/hero/basic-client/saved.html', 'utf-8');
-const document = DetachedDOM.load(html);
+const fragment = DetachedDOM.loadFragment(html);
```
[the remark this answers is not in the document]
Comment (calebjclark) [reply docs/main/BasicClient/DetachedDOM.md]: Fixed

**Pull request merged**